File: [C]hashmap/[C]hashmap/128-longest-consecutive-sequence.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        hashset = set(nums)

        res = 0
        for num in hashset:
            if num - 1 in hashset:
                continue
            if num + res not in hashset:
                continue
            cur = num
            while cur + 1 in hashset:
                cur = cur + 1
            res = max(res, cur - num + 1)
        return res
    
# time O(n)
# sapce O(n)
# using hashmap and store val and hashset and pruning
'''
1. find every sequence's start point
'''

File: [C]hashmap/[C]hashmap/128-longest-consecutive-sequence.java
class Solution {
    public int longestConsecutive(int[] nums) {
        HashSet<Integer> set = new HashSet<>();
        for (int num: nums) {
            set.add(num);
        }
        int res = 0;
        for (int num: nums) {
            if (set.contains(num - 1)) {
                continue;
            }
            int left = num;
            int right = num;
            if (!set.contains(left + res)) {
                continue;
            }
            while (set.contains(right + 1)) {
                right += 1;
            }
            res = Math.max(res, right - left + 1);
        }
        return res;
    }
}

// time O(n)
// sapce O(n)
// using hashmap and store val and hashset and pruning
/*
1. find every sequence's start point
*/